packages/adminrouter/extra/src/test-harness/tests/test_master.py

```diff
@@ -1,183 +1,368 @@
+             "http://127.0.0.1/foo/bar",
+             "http://127.0.0.1/service/scheduler-alwaysthere/foo/bar"),
+         LocationRewriteExpectation(
+            "/foo/bar",
+            "http://127.0.0.1/service/scheduler-alwaysthere/foo/bar"),
+         ]),
+    LocationHeaderRewriteTestConfig(
+        "http://127.0.0.1:8181",
+        "/exhibitor/v1/ui/index.html",
+        [LocationRewriteExpectation(
+            "http://127.0.0.1/exhibitor/v1/ui/index.html",
+            "http://127.0.0.1/exhibitor/exhibitor/v1/ui/index.html"),
+         ]),
+]
+
+location_header_test_params = [
+    (x.endpoint_id, x.basepath, y.location_set, y.location_expected)
+    for x in location_header_rewrite_test_config for y in x.redirect_testscases]
+
+redirected_paths = [
+    '/service/scheduler-alwaysthere',
+    '/exhibitor',
+    '/system/v1/metrics',
+    '/system/v1/logs/v1',
+]
+
+
+class TestMaster:
+    @pytest.mark.parametrize("path,expected_upstream", correct_upstream_dest_endpoints)
+    def test_if_request_is_sent_to_correct_upstream(
+            self,
+            master_ar_process,
+            valid_user_header,
+            path,
+            expected_upstream):
 
         generic_correct_upstream_dest_test(master_ar_process,
                                            valid_user_header,
-                                           agent_prefix + "/foo/bar",
-                                           'http://127.0.0.2:15001',
+                                           path,
+                                           expected_upstream,
                                            )
 
-    @pytest.mark.parametrize("path_given,path_expected",
-                             [("/foo/bar", "/foo/bar"),
-                              ("", "/"),
-                              ("/", "/"),
-                              ])
-    def test_if_upstream_request_is_correct(self,
-                                            master_ar_process,
-                                            valid_user_header,
-                                            path_given,
-                                            path_expected):
-
-        prefixed_pg = agent_prefix + path_given
+    @pytest.mark.parametrize("path,pass_authheader", upstream_headers_endpoints)
+    def test_if_upstream_headers_are_correct(
+            self,
+            master_ar_process,
+            valid_user_header,
+            path,
+            pass_authheader,
+            ):
+
+        if pass_authheader is True:
+            generic_upstream_headers_verify_test(master_ar_process,
+                                                 valid_user_header,
+                                                 path,
+                                                 assert_headers=valid_user_header,
+                                                 )
+        elif pass_authheader is False:
+            generic_upstream_headers_verify_test(
+                master_ar_process,
```

Comment:
We should be consistent with args indentation formatting on function level so pick one style here :-)